internal/pkg/pkg.go

```diff
@@ -436,6 +466,117 @@ func Subpackages(rootPath string, matcher SubpackageMatcher, recursive bool) ([]
 	return paths, nil
 }
 
+// SubpackagesWithFs returns a slice of paths to any subpackages of the provided path.
+// The matcher parameter decides if all types of subpackages should be considered,
+// and the recursive parameter determines if only direct subpackages are
+// considered. All returned paths will be relative to the provided rootPath.
+// The top level package is not considered a subpackage. If the provided path
+// doesn't exist, an empty slice will be returned.
+// It operates on the input FileSystem as opposed to FileSystem on disk
+func SubpackagesWithFs(fsys filesys.FileSystem, rootPath string, matcher SubpackageMatcher, recursive bool) ([]string, error) {
+	const op errors.Op = "pkg.Subpackages"
+
+	if !fsys.Exists(rootPath) {
+		return []string{}, nil
+	}
+	packagePaths := make(map[string]bool)
+	if err := fsys.Walk(rootPath, func(path string, fi os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to read package %s: %w", rootPath, err)
+		}
+
+		// Ignore the root folder
+		if path == rootPath {
+			return nil
+		}
+
+		// For every folder, we check if it is a kpt package
+		if fi.IsDir() {
+			// Ignore anything inside the .git folder
+			// TODO: We eventually want to support user-defined ignore lists.
+			if fi.Name() == ".git" {
+				return filepath.SkipDir
+			}
+
+			// Check if the directory is the root of a kpt package
+			isPkg, err := IsPackageDirFs(fsys, path)
+			if err != nil {
+				return err
+			}
+
+			// If the path is the root of a subpackage, add the
+			// path to the slice and return SkipDir since we don't need to
+			// walk any deeper into the directory.
+			if isPkg {
+				kf, err := FSReadKptfile(fsys, path)
+				if err != nil {
+					return errors.E(op, types.UniquePath(path), err)
+				}
+				switch matcher {
+				case Local:
+					if kf.Upstream == nil {
+						packagePaths[path] = true
+					}
+				case Remote:
+					if kf.Upstream != nil {
+						packagePaths[path] = true
+					}
+				case All:
+					packagePaths[path] = true
+				default:
+
+				}
+				if !recursive {
+					return filepath.SkipDir
+				}
+				return nil
+			}
+		}
+		return nil
+	}); err != nil {
+		return []string{}, fmt.Errorf("failed to read package at %s: %w", rootPath, err)
+	}
+
+	paths := []string{}
+	for subPkgPath := range packagePaths {
+		relPath, err := filepath.Rel(rootPath, subPkgPath)
+		if err != nil {
+			return paths, fmt.Errorf("failed to find relative path for %s: %w", subPkgPath, err)
+		}
+		paths = append(paths, relPath)
+	}
+	return paths, nil
+}
+
+// IsPackageDirFs checks if the package with input path exists in the input FileSystem
+func IsPackageDirFs(fsys filesys.FileSystem, path string) (bool, error) {
```

Comment:
nit: naming; the Fs part of the name differs across functions:
IsPackageDirFs (camel case suffix)
FSReadKptfile (all caps prefix)

Reply:
Removed.